design/sketch.webidl

```diff
@@ -0,0 +1,532 @@
+
+typedef u32 WebGPULoadOpEnum;
+interface WebGPULoadOp {
+    const u32 CLEAR = 0;
+    const u32 LOAD = 1;
+};
+
+typedef u32 WebGPUStoreOpEnum;
+interface WebGPUStoreOp {
+    const u32 STORE = 0;
+};
+
+dictionary WebGPURenderPassAttachmentDescriptor {
+    WebGPUTextureView attachment;
+    WebGPULoadOp loadOp;
+    WebGPUStoreOp storeOp;
+}
+
+dictionary WebGPURenderPassDescriptor {
+    sequence<WebGPURenderPassAttachmentDescriptor> colorAttachments;
+    WebGPURenderPassAttachmentDescriptor depthStencilAttachmaent;
+};
+
+interface WebGPUCommandBuffer {
+};
+
+interface WebGPUCommandEncoder {
+    WebGPUCommandBuffer finishEncoding();
+
+    // Commands allowed outside of "passes"
+    void copyBufferToBuffer(WebGPUBuffer src,
+                            u32 srcOffset,
+                            WebGPUBuffer dst,
+                            u32 dstOffset,
+                            u32 size);
+    // TODO figure out all the arguments required for these
+    void copyBufferToTexture();
+    void copyTextureToBuffer();
+    void copyTextureToTexture();
+    void blit();
+
+    // Allowed in both compute and render passes
+    void setPushConstants(WebGPUShaderStageFlags stage,
+                          u32 offset,
+                          u32 count,
+                          ArrayBuffer data);
+    void setBindGroup(u32 index, WebGPUBindGroup bindGroup);
+    void setPipeline((WebGPUComputePipeline or WebGPURenderPipeline) pipeline);
+
+    // Compute pass commands
+    void beginComputePass();
+    void endComputePass();
+
+    void dispatch(u32 x, u32 y, u32 z);
+
+    // Render pass commands
+    void beginRenderPass(WebGPURenderPassDescriptor descriptor);
+    void endRenderPass();
+
+    void setBlendColor(float r, float g, float b, float a);
+    void setIndexBuffer(WebGPUBuffer buffer, u32 offset);
+    void setVertexBuffers(u32 startSlot, sequence<WebGPUBuffer> buffers, sequence<u32> offsets);
+
+    void draw(u32 vertexCount, u32 instanceCount, u32 firstVertex, u32 firstInstance);
+    void drawIndexed(u32 indexCount, u32 instanceCount, u32 firstIndex, u32 firstInstance);
+
+    // TODO add missing commands
+};
+
+// ****************************************************************************
+// OTHER (Fence, Queue SwapChain, Device)
+// ****************************************************************************
+
+// Fence
+interface WebGPUFence {
+    u32 getValue();
+};
+
+// Queue
+interface WebGPUQueue {
+    void submit(sequence<WebGPUCommandBuffer> buffers);
+    void updateFence(WebGPUFence fence, u32 value);
+};
+
+// SwapChain / RenderingContext
+dictionary WebGPUSwapChainDescriptor {
+    WebGPUTextureUsageFlags usage;
+    WebGPUTextureFormatEnum format;
+    u32 width;
+    u32 height;
+};
+
+interface WebGPUSwapChain {
+    void configure(WebGPUSwapChainDescriptor descriptor);
+    WebGPUTexture getNextTexture();
+    void present();
+};
+
+interface WebGPURenderingContext : WebGPUSwapChain {
```

Comment:
To be able to `canvas.getContext('webgpu')`?

Reply:
Yes

Reply:
Do we really need it? It's desirable (for Yandex Maps JS API at least) to be able to "drive@ several canvases from one "instance" of an engine (so we can create shaders and some ohter resources once and also not run into limitation on number of contexts). That calls for ability to create several swap chains, one for each canvas element. And there're some definitions for creating them in the sketch. Doesn't that introduce duplication?

Reply:
I'm not sure I understand, in this sketch the device is created out of thin air, and multiple canvases can be driven from the same device through their swapchain. A canvas rendering context is the only way to get a swapchain in this sketch.

Reply:
Oh, I think I've misunderstood the sketch at first. For some reason I missed that `WebGPURenderingContext` is just a canvas-specific swap chain (am I right?) and thought of it as a wrapper around a swap chain *and* a device.

Rafael make comment in hangouts discussion about constructors being preferable way to create entities for a Web API. I'd say having something like `WebGPUCanvasSwapChain` and being able to create it with `new`:

```js
const swapChain = new WebGPUCanvasSwapChain(canvas); // BTW, do we need device here?

renderFrameTo(swapChain.getNextTexture());
```

Reply:
> WebGPURenderingContext is just a canvas-specific swap chain
Yes, that's right.

> constructors being preferable way to create entities
I think this is one of the few places where this isn't true. A canvas can only have one RenderingContext. `getContext` is fallible and can return the same object multiple times. For example:
```
cvs.getContext('webgl') === cvs.getContext('webgl')
```
```
cvs.getContext('2d')
cvs.getContext('webgl') === null
```

Reply:
Would it be too difficult to add something like `getWebGPUSwapChain` to `HTMLCanvasElement`. I find calling it `Context` really confusing:(